### ISSUE TITLE
WAL fsync() and IOPS requirements

### ISSUE THREAD

**Issue by mdyring** [tendermint/tendermint#1589]
This is a feature request for lowering disk I/O requirements down the line. 

As this affects hardware requirements for non-validating nodes as well (sentries, which would ideally be used for scale-out) it would be nice to keep costs down by supporting lower spec'd hardware/non-provisioned IOPS on virtual instances.

Short version: A single WAL message is followed by fsync() today, which at minimum generates one write operation on disk. A msg is generated for each block part, vote, etc. so requirements will grow with network size and not just transaction volume.

It might be worthwhile to investigate if some of these writes can be batched (followed by fsync()) and still provide good enough guarantees.

Copied from Matrix chat:
```
idoor
Thanks for suggestions, I checked cpu, memory usages, they seem ok on my ubuntu machine, when I checked disk I/O with iotop command, the gaiad processes only took less then 1% of IO operation for disk write when syncing, but the other command called jbd2/sda1-8 took about 90% of IO operstions, not sure what that is, the gaiad process writes at 173K/s sometimes, right now I am resyncing my node on ubuntu, it is syncing slower than my other node on my second Mac, which it only took a short time to fully be synced, I will put the issue on github with steps I did and results.

mdyring
I imagine the disk is not SSD but spinning rust (magnetic)? It is likely maxed out due to random IO.. Not sure what would cause this, I am guessing the IAVL tree persisting data or the WAL. melekes any ideas?

Also give "iostat -dx 1" a try and get some numbers of read/write operations per second.

This is probably the same reason that certain EC2 instances are not able to catch up, as the disk IO is throttled.

I think this is worth investigating by developers though to see if disk IO can be reduced.
(so please open issue on github with details on IO activitity, then devs can evaluate if things need to be tweaked)

melekes
jbd is journaling block device

mdyring
yeah, so maxing out on fsync() basically as that would need to write to journal device?


melekes
consensus WAL does fsync for every msg now
not sure how we can optimise that without sacrificing guaranties
yeah, please open an issue

mdyring
consensus WAL does fsync for every msg now
yeah, so it would be one block == one msg? so in the optimistic case, one block = one write op?

i think i saw an issue on GH to remove WAL as it might not be needed. that might help, if it is indeed the case :-)

melekes
1 msg (block part, vote, ..) - 1 fsync
mempool WAL is not used right now
but consensus WAL is used and needed

mdyring
ah.. so 100 votes (with enough validators) means 100 WAL writes?

melekes
yes

mdyring
ok.. yeah, then SSD will be a requirement for sure.

melekes
yes

mdyring
is this also the case for nonvalidating nodes? i am thinking sentries.. running those on EC2 will need provisioned IOPS
ideally it should be "cheap" to run sentries for scaleout

melekes
don't we have some requirements for validator nodes already?

mdyring
not sure. i haven't seen them if that is the case.. adrianbrink ?
but anyway it is helpful to understand why there is so much WAL activity and why. i guess there might be ways to optimize it eventually.

melekes
> is this also the case for nonvalidating nodes?

if they are running consensus reactor, then yes.
```

**Comment by idoor88** [comment]
I just checked my node on ubuntu, last night it was about 100 blocks behind,  this morning it is about 2000 blocks behind, I just ran iostat -dx 1, this is the result:

cat iostat.txt 
Linux 4.13.0-38-generic  	05/18/2018 	_x86_64_	(8 CPU)

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.02     6.56    0.03   11.39     0.99    86.27    15.27     0.18   15.59   13.42   15.60  13.11  14.98

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    43.00    0.00   90.00     0.00   528.00    11.73     0.98   10.89    0.00   10.89  10.89  98.00

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    29.00    0.00   72.00     0.00   404.00    11.22     0.99   14.00    0.00   14.00  13.72  98.80

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    46.00    0.00   85.00     0.00   644.00    15.15     1.03   11.86    0.00   11.86  11.62  98.80

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    44.00    0.00   75.00     0.00   472.00    12.59     1.00   13.60    0.00   13.60  13.33 100.00

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    25.00    0.00   60.00     0.00   340.00    11.33     0.99   16.27    0.00   16.27  16.47  98.80

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    31.00    0.00   78.00     0.00   436.00    11.18     0.99   12.77    0.00   12.77  12.67  98.80

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    35.00    0.00   78.00     0.00   452.00    11.59     0.99   12.72    0.00   12.72  12.72  99.20

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    33.00    0.00   61.00     0.00   440.00    14.43     1.02   16.26    0.00   16.26  16.26  99.20

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    39.00    0.00   69.00     0.00   428.00    12.41     0.98   14.72    0.00   14.72  14.20  98.00

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    30.00    0.00   75.00     0.00   424.00    11.31     1.00   12.96    0.00   12.96  13.33 100.00

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    30.00    0.00   63.00     0.00   368.00    11.68     1.00   16.00    0.00   16.00  15.81  99.60

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    26.00    0.00   60.00     0.00   348.00    11.60     1.00   16.40    0.00   16.40  16.53  99.20

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    30.00    0.00   75.00     0.00   420.00    11.20     0.99   13.23    0.00   13.23  13.17  98.80

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    36.00    0.00   68.00     0.00   412.00    12.12     1.01   15.18    0.00   15.18  14.65  99.60

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    27.00    0.00   66.00     0.00   376.00    11.39     0.99   14.67    0.00   14.67  15.03  99.20

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    36.00    0.00   81.00     0.00   468.00    11.56     0.99   12.30    0.00   12.30  12.20  98.80

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    32.00    0.00   81.00     0.00   448.00    11.06     0.98   12.15    0.00   12.15  12.15  98.40

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    32.00    0.00   69.00     0.00   404.00    11.71     0.99   14.61    0.00   14.61  14.32  98.80

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    40.00    0.00   81.00     0.00   488.00    12.05     1.00   11.95    0.00   11.95  12.30  99.60

Device:         rrqm/s   wrqm/s     r/s     w/s    rkB/s    wkB/s avgrq-sz avgqu-sz   await r_await w_await  svctm  %util
sda               0.00    54.00    0.00   87.00     0.00   560.00    12.87     1.00   11.72    0.00   11.72  11.45  99.60

**Comment by idoor88** [comment]
Idoor

**Comment by idoor88** [comment]
By the way I am running the node with Desktop machine ubuntu 16.04 LTS 64 bit Disk 1.5TB Processor Intel® Core™ i7 CPU 870 @ 2.93GHz × 8, which is not in the cloud, it is in home network with Verizon FIOS

**Comment by mdyring** [comment]
Related to https://github.com/tendermint/tendermint/issues/1589

**Comment by ebuchman** [comment]
Maybe we should bring this back: https://github.com/tendermint/tendermint/issues/1428 @melekes 

Note the WAL is more of a concern for enterprise users or users of small networks. In a larger network, it is less critical, since it should be possible to fully recover lost information from the network. So we could offer a way to turn the WAL off completely, I think.

The more correct but more difficult solution is to not fsync with every message, but only after messages that will cause a state machine transition. That would probably be a big help here and maybe not that hard to implement.

There's more discussion about this from when we were testing it here: https://github.com/tendermint/tendermint/issues/573

Also linking issues 

- improvements: https://github.com/tendermint/tendermint/issues/785
- old notes and subtleties from CoreOS ETCD: https://github.com/tendermint/tendermint/issues/269

**Comment by idoor88** [comment]
One more test, from iotop command, when gaiad process is running, the jbd2/sda1-8 command shows 90% IO, and gaiad process shows IO speed of 50 - 160 K/s for DISK_WRITE, after I kills the gaiad process, the jbd2/sda1-8 process is also gone.

**Comment by ebuchman** [comment]
Ok, this was actually quite straight forward. Should be fixed in #1601

**Comment by ebuchman** [comment]
Released in 0.19.5 and in SDK v0.17.2 - please update for the testnet there :)
